### PR TITLE
feat: add modifyEnvironmentConfig hook

### PR DESCRIPTION
**Pull request by 9aoy**
## Summary

support modify environment config via `modifyEnvironmentConfig` hook.

Type:
```ts
type ModifyEnvironmentConfigUtils = {
  /** environment name. */
  name: string;
  /** Merge multiple Rsbuild environment config objects into one. */
  mergeEnvironmentConfig: (
    config: MergedEnvironmentConfig,
    ...configs: EnvironmentConfig[]
  ) => MergedEnvironmentConfig;
};

function ModifyEnvironmentConfig(
  callback: (
    config: MergedEnvironmentConfig,
    utils: ModifyEnvironmentConfigUtils,
  ) => MaybePromise<MergedEnvironmentConfig | void>,
): void;
```

Example: 

```ts
const myPlugin = () => ({
  setup: (api) => {
    api.modifyEnvironmentConfig((config, { name }) => {
      if (name !== 'web') {
          return;
       }
      config.html ||= {};
      config.html.title = 'My Default Title';
    });
  },
});
```
## Related Links

https://github.com/web-infra-dev/rsbuild/issues/2620

## Checklist

- [ ] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: packages/core/src/provider/initConfigs.ts [hunk 1]
```diff
@@ -107,11 +132,35 @@ export async function initRsbuildConfig({
...
     normalizeBaseConfig,
     context.rootPath,
   );
 
+  const {
+    dev: { assetPrefix, lazyCompilation, ...rsbuildSharedDev },
+    server,
+  } = normalizeBaseConfig;
+
+  for await (const [name, config] of Object.entries(mergedEnvironments)) {
```
Comment (chenjiahan) [inline packages/core/src/provider/initConfigs.ts:147]: unused await?

File: packages/core/src/provider/initConfigs.ts [hunk 2]
```diff
@@ -33,16 +34,33 @@ async function modifyRsbuildConfig(context: InternalContext) {
...
 
+async function modifyEnvironmentConfig(
+  context: InternalContext,
+  config: MergedEnvironmentConfig,
+  name: string,
+) {
+  logger.debug(`modify Rsbuild environment(${name}) config`);
+  const [modified] = await context.hooks.modifyEnvironmentConfig.call(config, {
+    name,
+    // @ts-expect-error
```
Comment (chenjiahan) [inline packages/core/src/provider/initConfigs.ts:45]: Add the reason for @ts-expect-error
```suggestion
    // @ts-expect-error some reason
```

File: packages/shared/src/types/config/index.ts [hunk 3]
```diff
@@ -97,6 +97,18 @@ export interface RsbuildConfig extends EnvironmentConfig {
   _privateMeta?: RsbuildConfigMeta;
 }
 
+export type MergedEnvironmentConfig = {
```
Comment (chenjiahan) [inline packages/shared/src/types/config/index.ts:100]: I think the only different of `MergedEnvironmentConfig` and `NormalizedEnvironmentConfig` is `DeepReadonly<>`?

Can we just merge the two types? We do not want the config types to become so complex.

File: packages/shared/src/types/hooks.ts [hunk 4]
```diff
@@ -86,11 +91,26 @@ export type ModifyRsbuildConfigUtils = {
   mergeRsbuildConfig: (...configs: RsbuildConfig[]) => RsbuildConfig;
 };
 
+export type ModifyEnvironmentConfigUtils = {
+  /** environment name. */
+  name: string;
+  /** Merge multiple Rsbuild environment config objects into one. */
+  mergeEnvironmentConfig: (
```
Comment (chenjiahan) [inline packages/shared/src/types/hooks.ts:98]: Can we just name it `mergeRsbuildConfig`?

The implementation is `mergeEnvironmentConfig`, and `environments` is just part of Rsbuild config.

File: packages/shared/src/types/config/index.ts (hunk 3 above)
Comment (9aoy) [reply packages/shared/src/types/config/index.ts:100]: `MergedEnvironmentConfig` only has configs that the user can modify, but `NormalizedEnvironmentConfig` has all configurations that are read-only.


`NormalizedEnvironmentConfig` = DeepReadonly<`MergedEnvironmentConfig`  + `SharedRsbuildConfig` (server / _privateMeta / ... )>

File: packages/shared/src/types/hooks.ts (hunk 4 above)
Comment (9aoy) [reply packages/shared/src/types/hooks.ts:98]: But user can't modify the shared rsbuild config(not works) via `mergeRsbuildConfig`, and merge environment config  via `mergeRsbuildConfig ` is not make sense.  

It likes below:

```ts
          api.modifyEnvironmentConfig((config, { name, mergeRsbuildConfig }) => {
            return mergeRsbuildConfig(config, {
              environments: {
               [name]: {
                  source: {
                    alias: {
                      '@common1': './src/common1',
                    },
                  },
                },
              },
            });
          });
```

Comment (chenjiahan) [reply packages/shared/src/types/hooks.ts:98]: get

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**